### ISSUE TITLE
Calbacks are not executed with the original traceId under load.

### ISSUE THREAD

**Issue by mkante** [openzipkin/zipkin-js#174]
We are a NodeJs shop and we are trying to instrument all our microservices (about 10). Our code base is huge, we are routing all outbound HTTP call to zipkin-instrumentation-request and memcache call goes through zipkin-instrumentation-memcached. So far our zipkin setup is fine UI+API. 
Unfortunately when we simulate concurrent request, callbacks are not executed with the original traceId. The problem is reproducable consistently. I know zinkinjs use CLS, previously we had issues with CLS but the team was unconfortable binding all async callbacks. 
1- Is zipkinjs made for only frontend?
2- Do we still have to bind all callbacks, even if we use zipkin-instrumentation-xxx?
3- Are we pushing the limit of the library?
4- whats our solutuons?

**Comment by eirslett** [comment]
> Is zipkinjs made for only frontend?

No, in fact it was made originally for backend/node.js. Browser support is a more recent addon.

> Do we still have to bind all callbacks, even if we use zipkin-instrumentation-xxx?

What do you mean by that?

> Are we pushing the limit of the library?

No, not at all. It is designed to handle production workloads.

> whats our solutuons?

There might be a bug in the library. If you're able to set up a minimum reproducible example script, then it's possible to figure out what's going wrong. I think @SimenB was also looking at replacing CLS with the new async hooks that shipped with Node.js 8 (if you are able to use Node 8)

**Comment by mkante** [comment]
By zipkin-instrumentation-xxx I mean any library providing zipkin instrumentation like zipkin-instrumentation-request, zipkin-instrumentation-memcached and more. Most of them are asynchronous but their callbacks are not bound to CLS context.
Unfortunaly we are not running Node.js 8

**Comment by adriancole** [comment]
sounds like we need tests to ensure that callbacks have the correct trace
identifiers in scope? is that right?

**Comment by mkante** [comment]
Correct

**Comment by eirslett** [comment]
If you look in this repo, you can see that we already have integration tests for the instrumentations - maybe you can modify one of the integration tests and see if you can make it fail, to reproduce the bug?

**Comment by mkante** [comment]
I setup two microservices, `server1` makes request to `server2`. The goal of the simulation is to prove that the async callbacks are executed with the wrong traceIDs (the callbacks are not bounded to CLS)

Here is my project structure 
```
/zipkin-test
  /package.json
  /service1.js
  /service2.js
```
The package.js content
```
{
  "name": "zipkin-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "express": "^4.16.2",
    "request": "^2.83.0",
    "request-promise": "^4.2.2",
    "zipkin": "^0.11.1",
    "zipkin-context-cls": "^0.11.0",
    "zipkin-instrumentation-express": "^0.11.1",
    "zipkin-instrumentation-request": "^0.11.1"
  }
}
```
The service1.js content
```
var Tracer = require('zipkin').Tracer;
var ConsoleRecorder = require('zipkin').ConsoleRecorder;
var CLSContext = require('zipkin-context-cls');
var request = require('request-promise');
var app = require('express')();
var wrapRequest = require('zipkin-instrumentation-request');
var zipkinMiddleware = require('zipkin-instrumentation-express').expressMiddleware;

var log = console.log;

// Create a tracer
var tracer = new Tracer({
  ctxImpl: new CLSContext('zipkin'),
  traceId128Bit: true,
  recorder: new ConsoleRecorder(function(){}),
  localServiceName: 'service-1'
});

log('Root TraceID: ', tracer.id.traceId);

var zipkinRequest = wrapRequest(request, { tracer: tracer });

// Install zipkin express middleware
app.use(zipkinMiddleware({ tracer: tracer }));

app.get('/service1', function(req, res){

  log('Incoming request trace ID: ' + tracer.id.traceId);
  zipkinRequest({
    url: 'http://localhost:3001/service2',
    timeout: 5000, // Drop the connection after 5 seconds
  }, function(err, response, body) {
    log('Callback TraceID: ', tracer.id.traceId);
    if (err) {
      res.status(500).send(err)
      return;
    }
    res.status(200).send(body)
  })
});

app.listen(3000);
```
The service2.js content
```
var request = require('request-promise');
var app = require('express')();
var log = console.log;

app.get('/service2', function(req, res){
  log('Pulling books for user: '+req.params.id);
  // Do not send the response, let the connect hanging
});

app.listen(3001);
```
Now start both services
`node service1.js`
`node service2.js`
Make a request to service1
`curl http://localhost:3000/service1`

After 5 second your request will timeout and `service1` output is like this
```
Root TraceID:  235eef58addf848dbaec83b8304bb365
Incoming request trace ID: 7c2db32eac44fa2f75bba4264b8fbd92
Callback TraceID:  235eef58addf848dbaec83b8304bb365
Unhandled rejection RequestError: Error: ESOCKETTIMEDOUT
...
```
As you can see the callback traceID is wrong, the correct ID should be `7c2db32eac44fa2f75bba4264b8fbd92`.
I'm expecting the callback traceID to be `7c2db32eac44fa2f75bba4264b8fbd92`

**Comment by mkante** [comment]
If you update `service2` to send a response back, the callback traceID is correct

**Comment by mkante** [comment]
One of other our engineer found that the problem was coming from `request` module, the [response object get destroyed](https://github.com/request/request/blob/master/request.js#L1167)  on `ETSOCKETTIMEOUT` event. You can listen to the `request.on('response', callback)` and `request.on('error', callback)` but then your callbacks is executed outside of the tracer context.

Currently the work around I found is to manually bind `zipkinRequest` callback to the CLS context
```
...
log('Incoming request trace ID: ' + tracer.id.traceId);

var callback = tracer._ctxImpl._session.bind(callback);

zipkinRequest({
  url: 'http://localhost:3001/service2',
  timeout: 5000, // Drop the connection after 5 seconds
}, callback);
...
```

**Comment by Ghost---Shadow** [comment]
Are you (or any of your libraries) using async-await?

**Comment by jcchavezs** [comment]
@mkante is this still an issue, can you come up with a test for this?

**Comment by mkante** [comment]
I will take a look later
